src/dbnode/storage/index/block.go

```diff
@@ -443,63 +444,84 @@ func (b *block) queryWithSpan(
 	opts QueryOptions,
 	results DocumentResults,
 ) (bool, error) {
+	iter, err := b.QueryIter(ctx, query)
+	if err != nil {
+		return false, err
+	}
+	iterCloser := safeCloser{closable: iter}
+	defer func() {
+		_ = iterCloser.Close()
+		b.metrics.queryDocsMatched.RecordValue(float64(results.TotalDocsCount()))
+		b.metrics.querySeriesMatched.RecordValue(float64(results.Size()))
+	}()
+
+	if err := b.QueryWithIter(ctx, opts, iter, results, math.MaxInt64); err != nil {
```

Comment:
in a future PR, this code will move to the index worker. instead of `MaxInt64`, a configurable limit and a loop will be used.